hoodie-common/src/main/java/com/uber/hoodie/common/table/view/HoodieTableFileSystemView.java

```diff
@@ -186,47 +227,133 @@ private void writeObject(java.io.ObjectOutputStream out)
+    return getAllFileGroups(partitionPath)
+        .map(fileGroup -> {
+          FileSlice fileSlice = fileGroup.getLatestFileSlice().get();
+          // if the file-group is under compaction, pick the latest before compaction instant time.
+          if (isFileSliceAfterPendingCompaction(fileSlice)) {
+            String compactionInstantTime = fileIdToPendingCompactionInstantTime.get(fileSlice.getFileId());
+            return fileGroup.getLatestFileSliceBefore(compactionInstantTime);
+          }
+          return Optional.of(fileSlice);
+        })
         .map(Optional::get);
   }
 
+  /**
+   * Returns true if the file-group is under pending-compaction and the file-slice' baseInstant matches
+   * compaction Instant
+   * @param fileSlice File Slice
+   * @return
+   */
+  private boolean isFileSliceAfterPendingCompaction(FileSlice fileSlice) {
+    String compactionInstantTime = fileIdToPendingCompactionInstantTime.get(fileSlice.getFileId());
+    if ((null != compactionInstantTime) && fileSlice.getBaseInstantTime().equals(compactionInstantTime)) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * With concurrent compaction, it is possible to see partial/complete data-files due to inflight-compactions,
+   * Ignore those data-files
+   * @param fileSlice File Slice
+   * @return
+   */
+  private FileSlice unsetDataFileForFileSliceAfterPendingCompaction(FileSlice fileSlice) {
+    if (isFileSliceAfterPendingCompaction(fileSlice)) {
+      FileSlice transformed = new FileSlice(fileSlice.getBaseInstantTime(), fileSlice.getFileId());
+      fileSlice.getLogFiles().forEach(lf -> transformed.addLogFile(lf));
+      return transformed;
+    }
+    return fileSlice;
+  }
+
   @Override
   public Stream<FileSlice> getLatestFileSlicesBeforeOrOn(String partitionPath,
       String maxCommitTime) {
     return getAllFileGroups(partitionPath)
         .map(fileGroup -> fileGroup.getLatestFileSliceBeforeOrOn(maxCommitTime))
-        .filter(dataFileOpt -> dataFileOpt.isPresent())
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .map(this::unsetDataFileForFileSliceAfterPendingCompaction);
+  }
+
+  /**
+   * Helper to merge last 2 file-slices so that the last file-slice base-instant is considered outstanding compaction
+   * instant. These 2 file-slices do not have compaction done yet.
+   *
+   * @param lastSlice        Latest File slice for a file-group
+   * @param penultimateSlice Penultimate file slice for a file-group in commit timeline order
+   */
+  private static FileSlice mergeCompactionPendingFileSlices(FileSlice lastSlice, FileSlice penultimateSlice) {
+    FileSlice merged = new FileSlice(lastSlice.getBaseInstantTime(), lastSlice.getFileId());
+    if (penultimateSlice.getDataFile().isPresent()) {
+      merged.setDataFile(penultimateSlice.getDataFile().get());
+    }
+    // Add Log files from penultimate and last slices
+    lastSlice.getLogFiles().forEach(lf -> merged.addLogFile(lf));
```

Comment:
should nt we add penultimate logs first?

Reply:
This is probably needed for something like HoodieBloomIndex that relies on penultimate base file to do the indexing..

Reply:
Just another comment on log scanner being able to work across log files belonging to different base commits..

Reply:
Done.